[SwiftUI] Support media-related API functionality

rr-codes opened the pull request:
#### 00588182d1cf03e9414afa467d2266b7bc293317
<pre>
[SwiftUI] Support media-related API functionality
<a href="https://bugs.webkit.org/show_bug.cgi?id=284651">https://bugs.webkit.org/show_bug.cgi?id=284651</a>
<a href="https://rdar.apple.com/141109070">rdar://141109070</a>

Reviewed by Aditya Keerthi.

Add functionality to support controlling media interaction to WebPage and WebView:

* Adds an observable `fullscreenState` to WebPage which is analogous to the one on WKWebView.
* Adds some functions to WebPage which are analogous to their WebPage counterparts, such as `setAllMediaPlaybackSuspended`
* Adds several properties and functions to WebPage to be able to observe and control camera and microphone capture state.
* Adds support for customizing when permission prompts are displayed.

* Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift:
(EnvironmentValues.webViewCameraCaptureState):
(EnvironmentValues.webViewMicrophoneCaptureState):
(View.webViewCameraCaptureState(_:)):
(View.webViewMicrophoneCaptureState(_:)):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(fullscreenState):
(pauseAllMediaPlayback):
(mediaPlaybackState):
(setAllMediaPlaybackSuspended(_:)):
(closeAllMediaPresentations):
(WebPage_v0.handlesURLScheme(_:)): Deleted.
(WebPage_v0.backForwardList): Deleted.
(WebPage_v0.url): Deleted.
(WebPage_v0.title): Deleted.
(WebPage_v0.estimatedProgress): Deleted.
(WebPage_v0.isLoading): Deleted.
(WebPage_v0.serverTrust): Deleted.
(WebPage_v0.hasOnlySecureContent): Deleted.
(WebPage_v0.isWritingToolsActive): Deleted.
(WebPage_v0.mediaType): Deleted.
(WebPage_v0.customUserAgent): Deleted.
(WebPage_v0.isInspectable): Deleted.
(WebPage_v0.backingWebView): Deleted.
(WebPage_v0.load(_:)): Deleted.
(WebPage_v0.load(_:mimeType:characterEncoding:baseURL:)): Deleted.
(WebPage_v0.load(_:baseURL:)): Deleted.
(WebPage_v0.load(fileURL:allowingReadAccessTo:)): Deleted.
(WebPage_v0.load(fileRequest:allowingReadAccessTo:)): Deleted.
(WebPage_v0.load(simulatedRequest:response:responseData:)): Deleted.
(WebPage_v0.load(simulatedRequest:responseHTML:)): Deleted.
(WebPage_v0.reload(_:)): Deleted.
(WebPage_v0.stopLoading): Deleted.
(WebPage_v0.callAsyncJavaScript(_:arguments:in:contentWorld:)): Deleted.
(WebPage_v0.snapshot(_:)): Deleted.
(WebPage_v0.pdf(_:)): Deleted.
(WebPage_v0.webArchiveData): Deleted.
* Source/WebKit/UIProcess/API/Swift/WebView.swift:
(WebViewRepresentable.makePlatformView(_:)):
(WebViewRepresentable.updatePlatformView(_:context:)):
(WebViewRepresentable.makeCoordinator):
(Coordinator.cameraCaptureState):
(Coordinator.microphoneCaptureState):
(Coordinator.observations):
(Coordinator.owner):
(Coordinator.updateObservations):
* Tools/SwiftBrowser/Resources/iOS/Info.plist:
* Tools/SwiftBrowser/Resources/mac/Info.plist:
* Tools/SwiftBrowser/Source/ContentView.swift:
(LabelConfiguration.captureState):
(LabelConfiguration.body):
(PrincipalToolbarGroup.cameraCaptureState):
(PrincipalToolbarGroup.microphoneCaptureState):
(PrincipalToolbarGroup.body):
(ContentView.cameraCaptureState):
(ContentView.microphoneCaptureState):
(ContentView.body):

Canonical link: <a href="https://commits.webkit.org/288251@main">https://commits.webkit.org/288251@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d4dd97237ef34af4459649ec85f081ea5e94075a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82250 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1914 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36315 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87382 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33311 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84355 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1985 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9795 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64103 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21851 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85319 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1380 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74840 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44381 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1282 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29021 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32352 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72582 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29643 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88738 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9556 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6817 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72495 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9782 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70654 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71713 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15847 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/14864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/909 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12764 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9509 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15024 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9383 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12849 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11152 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+Configuration.swift:
```diff
@@ -63,6 +65,26 @@ extension WebPage_v0 {
+            case mediaCapture(WKMediaCaptureType)
+        }
+
+        let decision: (Permission, WebPage_v0.FrameInfo, WKSecurityOrigin) async -> WKPermissionDecision
```
For a block I’d probably give it a more action-y name, like `decisionHandler` to clarify it isn’t a decision in itself. And then ditto as the argument name.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebView.swift:
```diff
@@ -115,6 +120,18 @@ fileprivate struct WebViewRepresentable {
+    }
+
+    func makeCoordinator() -> Coordinator {
+        Coordinator(cameraCaptureState: cameraCaptureState, microphoneCaptureState: microphoneCaptureState)
```
These bindings can refresh, so you might end up with stale values if you only take them in during object creation. (Similar to: https://github.com/WebKit/WebKit/pull/37933#discussion_r1885450375, but more like `webViewCameraCaptureState(Bool.random() ? $binding1 : $binding2)` in construction)

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebView.swift:
```diff
@@ -140,4 +157,67 @@ extension WebViewRepresentable: NSViewRepresentable {
 #endif
 
+@MainActor
+private final class Coordinator {
```
Since we’re declaring this top level, is there a more descriptive name it could get?

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebView.swift:
```diff
@@ -140,4 +157,67 @@ extension WebViewRepresentable: NSViewRepresentable {
+        }
+    }
+
+    private func createObservation<Value>(on webView: WKWebView, keyPath: KeyPath<Coordinator, Binding<Value>?>, backedBy backingKeyPath: KeyPath<WKWebView, Value>) -> NSKeyValueObservation where Value: Equatable {
```
Might not hurt to line break since it’s got a lot going on.
```suggestion
    private func createObservation<Value>(
        on webView: WKWebView,
        keyPath: KeyPath<Coordinator, Binding<Value>?>,
        backedBy backingKeyPath: KeyPath<WKWebView, Value>
    ) -> NSKeyValueObservation where Value: Equatable {
```

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebView.swift:
```diff
@@ -140,4 +157,67 @@ extension WebViewRepresentable: NSViewRepresentable {
+        } else {
+            for observation in observations {
+                observation.invalidate()
+            }
```
`observations = []` after this?

axiixc commented on Tools/SwiftBrowser/Source/ContentView.swift:
```diff
@@ -55,7 +55,97 @@ fileprivate struct ToolbarBackForwardMenuView: View {
+            }
+
+        case .muted:
+            Button {
```
If this were for real, I’d suggest some fun stuff you can do with SF Symbols to make this a nicer state transition, but for a test app this is fine :P

rr-codes replied:
you've piqued my curiosity now 😛

axiixc commented on Tools/SwiftBrowser/Source/ContentView.swift:
```diff
@@ -55,7 +55,97 @@ fileprivate struct ToolbarBackForwardMenuView: View {
+                    viewModel.page.stopLoading()
+                } label: {
+                    Image(systemName: "xmark")
+                }
```
This is normally `⌘.` for a shortcut.

axiixc commented on Tools/SwiftBrowser/Source/ContentView.swift:
```diff
@@ -55,7 +55,97 @@ fileprivate struct ToolbarBackForwardMenuView: View {
+    }
+}
+
+private struct PrincipalToolbarGroup: View {
```
You could conform this to `ToolbarContent, DynamicProperty` instead, and then use it directly in the `toolbar` below (but this is also fine).

rr-codes replied:
how would I do this? When I try simply just conforming it and nothing else, I get these errors. When I try to make `body` be `some ToolbarContent`, I get even more errors.


<img width="640" alt="image" src="https://github.com/user-attachments/assets/420b6656-0fed-4e98-bcac-f120d5ff954d" />

axiixc commented on Tools/SwiftBrowser/Source/SettingsView.swift:
```diff
@@ -22,23 +22,58 @@
+                    permissionDecision: $mediaCaptureAuthorization,
+                    label: "Allow sites to access camera:"
+                )
+                .padding(.top)
```
Do you want the padding inside the section? Won’t that just make this row extra tall?

rr-codes replied:
I want the sections to have padding in between themselves. When I add padding to the Section directly, it for some reason causes padding for each individual item of the section

axiixc commented on Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift:
```diff
@@ -63,6 +67,16 @@ extension View {
     }
+
+    @_spi(Private)
+    public func webViewCameraCaptureState(_ value: Binding<WKMediaCaptureState>) -> some View {
```
I think I’m missing what ensures that the bindings are updated promptly for their initial layout pass. While I understand using a view modifier for these, since they’re very UI-bound, this state still suffers from the denormalized ownership problem that we avoided by placing stuff like `title` directly on `WebPage`. Basically, by forcing clients to own some canonical `@State` here, while WebKit also has it’s own definition, we set ourselves up to need to ensure these stay in sync. This is doable, but less free than a `var { get set }` on the `WebPage` object would be.

We also still need to work out how “re-parenting” of a `WebPage` will work with this API, which would be a similar synchronization event we’d want to test for. Ex:
```swift
struct Example : View {
  var page: WebPage
  @State var captureState1: WKMediaCaptureState = .none
  @State var captureState2: WKMediaCaptureState = .none

  var body: some View {
    if Bool.random() { // Extremely contrived example
      WebView(page).webViewCameraCaptureState($captureState1)
    } else {
      // Separate view identity here
      WebView(page).webViewCameraCaptureState($captureState2)
    }
  }
}
```

rr-codes committed:
[SwiftUI] Support media-related API functionality
https://bugs.webkit.org/show_bug.cgi?id=284651
rdar://141109070

Reviewed by Aditya Keerthi.

Add functionality to support controlling media interaction to WebPage and WebView:

* Adds an observable `fullscreenState` to WebPage which is analogous to the one on WKWebView.
* Adds some functions to WebPage which are analogous to their WebPage counterparts, such as `setAllMediaPlaybackSuspended`
* Adds several properties and functions to WebPage to be able to observe and control camera and microphone capture state.
* Adds support for customizing when permission prompts are displayed.

* Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift:
(EnvironmentValues.webViewCameraCaptureState):
(EnvironmentValues.webViewMicrophoneCaptureState):
(View.webViewCameraCaptureState(_:)):
(View.webViewMicrophoneCaptureState(_:)):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(fullscreenState):
(pauseAllMediaPlayback):
(mediaPlaybackState):
(setAllMediaPlaybackSuspended(_:)):
(closeAllMediaPresentations):
(WebPage_v0.handlesURLScheme(_:)): Deleted.
(WebPage_v0.backForwardList): Deleted.
(WebPage_v0.url): Deleted.
(WebPage_v0.title): Deleted.
(WebPage_v0.estimatedProgress): Deleted.
(WebPage_v0.isLoading): Deleted.
(WebPage_v0.serverTrust): Deleted.
(WebPage_v0.hasOnlySecureContent): Deleted.
(WebPage_v0.isWritingToolsActive): Deleted.
(WebPage_v0.mediaType): Deleted.
(WebPage_v0.customUserAgent): Deleted.
(WebPage_v0.isInspectable): Deleted.
(WebPage_v0.backingWebView): Deleted.
(WebPage_v0.load(_:)): Deleted.
(WebPage_v0.load(_:mimeType:characterEncoding:baseURL:)): Deleted.
(WebPage_v0.load(_:baseURL:)): Deleted.
(WebPage_v0.load(fileURL:allowingReadAccessTo:)): Deleted.
(WebPage_v0.load(fileRequest:allowingReadAccessTo:)): Deleted.
(WebPage_v0.load(simulatedRequest:response:responseData:)): Deleted.
(WebPage_v0.load(simulatedRequest:responseHTML:)): Deleted.
(WebPage_v0.reload(_:)): Deleted.
(WebPage_v0.stopLoading): Deleted.
(WebPage_v0.callAsyncJavaScript(_:arguments:in:contentWorld:)): Deleted.
(WebPage_v0.snapshot(_:)): Deleted.
(WebPage_v0.pdf(_:)): Deleted.
(WebPage_v0.webArchiveData): Deleted.
* Source/WebKit/UIProcess/API/Swift/WebView.swift:
(WebViewRepresentable.makePlatformView(_:)):
(WebViewRepresentable.updatePlatformView(_:context:)):
(WebViewRepresentable.makeCoordinator):
(Coordinator.cameraCaptureState):
(Coordinator.microphoneCaptureState):
(Coordinator.observations):
(Coordinator.owner):
(Coordinator.updateObservations):
* Tools/SwiftBrowser/Resources/iOS/Info.plist:
* Tools/SwiftBrowser/Resources/mac/Info.plist:
* Tools/SwiftBrowser/Source/ContentView.swift:
(LabelConfiguration.captureState):
(LabelConfiguration.body):
(PrincipalToolbarGroup.cameraCaptureState):
(PrincipalToolbarGroup.microphoneCaptureState):
(PrincipalToolbarGroup.body):
(ContentView.cameraCaptureState):
(ContentView.microphoneCaptureState):
(ContentView.body):

Canonical link: https://commits.webkit.org/288251@main
